rfcs/0088-commit-message-guideline.md

```diff
@@ -0,0 +1,187 @@
+messages and so improve the current situation regarding the TVM commit logs.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Commit Message Guideline
+
+Apache TVM uses the Github (GH) platform for patch submission and code review
+via Pull Requests (PRs). The final commit (title and body) that is merged into
+the Apache TVM main tree is composed of the PR's title and body and must be kept
+updated and reflecting the new changes in the code as per the reviews and
+discussions.
+
+Although these guidelines apply essentially to the PRs’ title and body messages,
+because GH auto-generates the PR’s title and body from the commits on a given
+branch, it’s recommended to follow these guidelines right from the beginning,
+when preparing commits in general to be submitted to the Apache TVM project.
+This will ease the creation of a new PR, avoiding rework, and also will help the
+review.
+
+The rules below will help to achieve uniformity that has several benefits, both
+for review and for the code base maintenance as a whole, helping you to write
+commit messages with a good quality suitable for the Apache TVM project,
+allowing fast log searches, bisecting, and so on.
+
+_PR/commit title_:
+
+* Guarantee a title exists (enforced);
+* Don’t use Github usernames in the title, like @username (enforced);
+* Check if a tag should be present as a hint about what component(s) of the code
+  the commits “touch”. For example [BugFix], [CI], [microTVM], and [TVMC]. Tags
+  go between square brackets and appear first in the title. If more than one tag
+  exist, multiple brackets should be used, like [BugFix][CI]. The case
+  recommended for tags, in geral, is the upper camel case. For example, prefer
+  the forms [Fix], [BugFix], and [Docker] instead of [fix], [bug_fix], and
+  [docker]. Acronyms should be kept as such so, for example, use [CI] and [TVMC]
+  instead of [ci] and [tvmc]. Tags help reviewers to identify the PRs they
+  can/want to review and also help the release folks when generating the release
+  notes (enforced);
+* Use an imperative mood. Avoid titles like “Added operator X” and “Updated
+  image Y in the CI”, instead use the forms “Add feature X” and “Update image Y
+  in the CI” instead;
+* Observe proper use of caps at the beginning (uppercase for the first letter)
+  and for acronyms, like, for instance, TVM, FVP, OpenCL. Hence instead of
+  “fix tvm use of opencl library”, write it as “Fix TVM use of OpenCL library”;
+* No period at the end of the title is necessary.
+
+_PR/commit body_:
+
+* Guarantee a body exists (enforced);
+* Don’t use Github usernames in body text, like @username (enforced);
+* Avoid “bullet” commit message bodies: “bullet” commit message bodies are not
+  bad per se, but “bullet” commit messages without any description or
+  explanation is likely as bad as commits without any description, rationale,
+  or explanation in the body.
```

Comment:
To clarify, does the "bullet commit message" mean the messages generated by GitHub squash merge? e.g.

```
This is a PR title (#123)
* upd
* save
* fix
* done
```

On the contrary, this is encouraged to have commit messages like:

```
This is PR title (#123)

This PR introduces the following changes:
* Add more flexibility to `Schedule.cache_read`
* Fix a bug in the original `cache_read` implementation with a regression test added
* Polish the doc to make it more readable
```

Reply:
To reinforce the original intent of the RFC, the bullet commit guidance is merely a recommendation. I suppose that as in any text, the `*` signed would be used where you need an unordered list, and that's ok IMHO.

My expectation is that if we get commits described with meaningful phrases stating the work done, and it happens to be a bullet list, it is still a step forward from what we have now.

Reply:
That makes a lot of sense to me! Thank you for the clarification!